### ISSUE TITLE
Adding an enum field does not generate a migration

### ISSUE THREAD

**Issue by pezholio** [typeorm/typeorm#8568]
## Issue Description

When I add an enum item to an existing enum field on an Entity and run `migration:generate`, a blank migration is created

### Expected Behavior

I expect a migration to be created to add add the field to the Postgres database

### Actual Behavior

An empty migration is created

### Steps to Reproduce

1. Add an entity with an enum field:

```typescript
export enum Status {
  Live = 'live',
  Draft = 'draft',
  Archived = 'archived',
}

@Entity({ name: 'something' })
export class Something {

  @Column({
    type: 'enum',
    enum: Status,
    default: Status.Draft,
  })
  status: OrganisationVersionStatus;
}
```

2. Generate and run the migrations
3. Add another enum field:

```typescript
export enum Status {
  Live = 'live',
  Draft = 'draft',
  Archived = 'archived',
  Unconfirmed = 'unconfirmed',
}

@Entity({ name: 'something' })
export class Something {

  @Column({
    type: 'enum',
    enum: Status,
    default: Status.Draft,
  })
  status: OrganisationVersionStatus;
}
```
4. Generate the migrations
5. Result:

```typescript
export class AddUnconfirmedStatusToSomething643119746947 implements MigrationInterface
{
  name = 'AddUnconfirmedStatusToSomething1643119746947';

  public async up(queryRunner: QueryRunner): Promise<void> {}

  public async down(queryRunner: QueryRunner): Promise<void> {}
}
```

### My Environment

| Dependency          | Version  |
| ---                 | ---      |
| Operating System    |  Mac OS        |
| Node.js version     | 16.13.0  | 
| Typescript version  | 4.4.4  | 
| TypeORM version     | 0.2.40  | 

### Relevant Database Driver(s)

| DB Type              | Reproducible  |
| ---                  | ---           |
| `aurora-data-api`    | no            |
| `aurora-data-api-pg` | no            |
| `better-sqlite3`     | no            |
| `cockroachdb`        | no            |
| `cordova`            | no            |
| `expo`               | no            |
| `mongodb`            | no            |
| `mysql`              | no            |
| `nativescript`       | no            |
| `oracle`             | no            |
| `postgres`           | yes            |
| `react-native`       | no            |
| `sap`                | no            |
| `sqlite`             | no            |
| `sqlite-abstract`    | no            |
| `sqljs`              | no            |
| `sqlserver`          | no            |

### Are you willing to resolve this issue by submitting a Pull Request?

 - ✖️ Yes, I have the time, and I know how to start.
 - ✖️ Yes, I have the time, but I don't know how to start. I would need guidance.
 - ✖️ No, I don’t have the time, but I can support (using donations) development.
 -  ✅ No, I don’t have the time and I’m okay to wait for the community / maintainers to resolve this issue.

**Comment by pezholio** [comment]
@Ginden I'm using NestJS, but I think this is the pertinent bit of my config:

```
{
    type: 'postgres',
    url: process.env.DATABASE_URL,
    entities: ['./dist/**/*.entity.js'],
    synchronize: false,
    migrations: ['./dist/db/migrate/*.js'],
    migrationsRun: true,
    dropSchema: process.env.NODE_ENV == 'test',
    cli: {
      migrationsDir: 'src/db/migrate',
    },
    extra: {
      ssl:
        process.env.NODE_ENV === 'production'
          ? { rejectUnauthorized: false }
          : false,
    },
  }
```

**Comment by pezholio** [comment]
And yep. I ran the compiler first, because I made a change to the default, and this was reflected in the generated migration 👍

**Comment by Ahmad-RW** [comment]
can you try and set the type of `status` to string and generate the migration.

**Comment by pezholio** [comment]
@Ahmad-RW I get this error:

```
DataTypeNotSupportedError: Data type "string" in "OrganisationVersion.status" is not supported by "postgres" database.
```

**Comment by Ahmad-RW** [comment]
> @Ahmad-RW I get this error:
> 
> ```
> DataTypeNotSupportedError: Data type "string" in "Something.status" is not supported by "postgres" database.
> ```
sorry I meant something like this :
```typescript
export enum Status {
  Live = 'live',
  Draft = 'draft',
  Archived = 'archived',
}

@Entity({ name: 'something' })
export class Something {

  @Column({
    type: 'enum',
    enum: Status,
    default: Status.Draft,
  })
  status: string;
}
```

**Comment by pezholio** [comment]
@Ahmad-RW Ah ha! That's got it! I now see:

```import { MigrationInterface, QueryRunner } from 'typeorm';

export class Something1643195466406 implements MigrationInterface {
  name = 'Something1643195466406';

  public async up(queryRunner: QueryRunner): Promise<void> {
    await queryRunner.query(
      `ALTER TYPE "public"."organisationVersions_status_enum" RENAME TO "organisationVersions_status_enum_old"`,
    );
    await queryRunner.query(
      `CREATE TYPE "public"."organisationVersions_status_enum" AS ENUM('live', 'draft', 'archived', 'awaiting')`,
    );
    await queryRunner.query(
      `ALTER TABLE "organisationVersions" ALTER COLUMN "status" DROP DEFAULT`,
    );
    await queryRunner.query(
      `ALTER TABLE "organisationVersions" ALTER COLUMN "status" TYPE "public"."organisationVersions_status_enum" USING "status"::"text"::"public"."organisationVersions_status_enum"`,
    );
    await queryRunner.query(
      `ALTER TABLE "organisationVersions" ALTER COLUMN "status" SET DEFAULT 'draft'`,
    );
    await queryRunner.query(
      `DROP TYPE "public"."organisationVersions_status_enum_old"`,
    );
  }

  public async down(queryRunner: QueryRunner): Promise<void> {
    await queryRunner.query(
      `CREATE TYPE "public"."organisationVersions_status_enum_old" AS ENUM('live', 'draft', 'archived')`,
    );
    await queryRunner.query(
      `ALTER TABLE "organisationVersions" ALTER COLUMN "status" DROP DEFAULT`,
    );
    await queryRunner.query(
      `ALTER TABLE "organisationVersions" ALTER COLUMN "status" TYPE "public"."organisationVersions_status_enum_old" USING "status"::"text"::"public"."organisationVersions_status_enum_old"`,
    );
    await queryRunner.query(
      `ALTER TABLE "organisationVersions" ALTER COLUMN "status" SET DEFAULT 'draft'`,
    );
    await queryRunner.query(
      `DROP TYPE "public"."organisationVersions_status_enum"`,
    );
    await queryRunner.query(
      `ALTER TYPE "public"."organisationVersions_status_enum_old" RENAME TO "organisationVersions_status_enum"`,
    );
  }
}


```

**Comment by Ahmad-RW** [comment]
> @Ahmad-RW Ah ha! That's got it! I now see:
> 
> ```ts
> import { MigrationInterface, QueryRunner } from 'typeorm';
> 
> export class Something1643195466406 implements MigrationInterface {
>   name = 'Something1643195466406';
> 
>   public async up(queryRunner: QueryRunner): Promise<void> {
>     await queryRunner.query(
>       `ALTER TYPE "public"."organisationVersions_status_enum" RENAME TO "organisationVersions_status_enum_old"`,
>     );
>     await queryRunner.query(
>       `CREATE TYPE "public"."organisationVersions_status_enum" AS ENUM('live', 'draft', 'archived', 'awaiting')`,
>     );
>     await queryRunner.query(
>       `ALTER TABLE "organisationVersions" ALTER COLUMN "status" DROP DEFAULT`,
>     );
>     await queryRunner.query(
>       `ALTER TABLE "organisationVersions" ALTER COLUMN "status" TYPE "public"."organisationVersions_status_enum" USING "status"::"text"::"public"."organisationVersions_status_enum"`,
>     );
>     await queryRunner.query(
>       `ALTER TABLE "organisationVersions" ALTER COLUMN "status" SET DEFAULT 'draft'`,
>     );
>     await queryRunner.query(
>       `DROP TYPE "public"."organisationVersions_status_enum_old"`,
>     );
>   }
> 
>   public async down(queryRunner: QueryRunner): Promise<void> {
>     await queryRunner.query(
>       `CREATE TYPE "public"."organisationVersions_status_enum_old" AS ENUM('live', 'draft', 'archived')`,
>     );
>     await queryRunner.query(
>       `ALTER TABLE "organisationVersions" ALTER COLUMN "status" DROP DEFAULT`,
>     );
>     await queryRunner.query(
>       `ALTER TABLE "organisationVersions" ALTER COLUMN "status" TYPE "public"."organisationVersions_status_enum_old" USING "status"::"text"::"public"."organisationVersions_status_enum_old"`,
>     );
>     await queryRunner.query(
>       `ALTER TABLE "organisationVersions" ALTER COLUMN "status" SET DEFAULT 'draft'`,
>     );
>     await queryRunner.query(
>       `DROP TYPE "public"."organisationVersions_status_enum"`,
>     );
>     await queryRunner.query(
>       `ALTER TYPE "public"."organisationVersions_status_enum_old" RENAME TO "organisationVersions_status_enum"`,
>     );
>   }
> }
> ```

Nice. Is that what you wanted? I think enums in Postgres are not very good. If I recall correctly, enums in Postgres aren't very good. if you want to remove elements to it. You'll have to drop the enum and recreate it without the removed element. But it's up to you to review the pros and cons :).
if your issue is resolved please close the issue.
thanks.

**Comment by pezholio** [comment]
Yeah, that works now, but I would expect the migrations to be created if the column has the enum type set too, so I still think this is a bug?

**Comment by Ahmad-RW** [comment]
can you share your entity after the latest change you made?

**Comment by alex-demchenko** [comment]
Issue still exists
Version: 0.3.17

**Comment by softzer0** [comment]
Yep, can confirm this still being a bug, and we can't put `string` instead of Enum for the columns because of type mismatch with equivalent DTOs. 😞

**Comment by adriendomoison** [comment]
This happened to me because I was working with `sync: true` during my initial phase of development, and when I dropped the tables, **I forgot to drop the enums** as well so they didn't generate. This is probably why adding `_enum` worked for @nikdale but there is most likely no need to. Hope this helps someone!

**Comment by nikdale** [comment]
> This happened to me because I was working with `sync: true` during my initial phase of development, and when I dropped the tables to generate the migrations, **I forgot to drop the enums** as well so they didn't generate. This might be why adding `_enum` worked for @nikdale but there is most likely no need to. Hope this helps someone!

Could be that people only did dropped the tables but forgot that they have to drop whole schema, but if anyone has this issue, try adding the `_enum` part and if it works, you probably have to drop whole schema, remove added `_enum` part and it should create it in new migrations.

**Comment by ZahitAtrato** [comment]
If you're lookin a way to create Enum files usign the TypeORM api in your migrations:

`const table = new Table({
	name: 'Something',
	columns: [
		{
			name: 'status',
			type: 'enum',
			length: "'live', 'draft', 'archived'"
		},
	]
});
await queryRunner.createTable(table, true);`

**Comment by dizonb** [comment]
I have a similar experience, wonder why Datatypes are UNKNOWN. I'm using HeidiSQL.

```
export enum ArticleType {
  NEWS = 'news',
  NEWS_RELEASE = 'news_release',
  COLUMN = 'column',
  CASE = 'case',
  OTHER = 'other',
}

export enum ArticleStatus {
  DRAFT = 'draft',
  PUBLISHED = 'published',
  ARCHIVED = 'archived',
}
```

![Image](https://github.com/user-attachments/assets/f9fdda95-e43c-4f94-81dc-89bc8c7c8044)

```
@Column({
    type: 'enum',
    enum: ArticleType,
    default: ArticleType.OTHER,
  })
  articleType: ArticleType;

@Column({
    type: 'enum',
    enum: ArticleStatus,
    default: ArticleStatus.DRAFT,
  })
  status: ArticleStatus;
```

I tried above suggestions, however, no luck.

**Comment by dizonb** [comment]
I can do migration with this, but wonder why Datatype for enums are UNKNOWN. I'm using HeidiSQL instead of pgadmin4.

```
export enum ArticleType {
  NEWS = 'news',
  NEWS_RELEASE = 'news_release',
  COLUMN = 'column',
  CASE = 'case',
  OTHER = 'other',
}

export enum ArticleStatus {
  DRAFT = 'draft',
  PUBLISHED = 'published',
  ARCHIVED = 'archived',
}
```


```
@Column({
    type: 'enum',
    enum: ArticleType,
    default: ArticleType.OTHER,
  })
  articleType: ArticleType;

@Column({
    type: 'enum',
    enum: ArticleStatus,
    default: ArticleStatus.DRAFT,
  })
  status: ArticleStatus;
```

![Image](https://github.com/user-attachments/assets/f9fdda95-e43c-4f94-81dc-89bc8c7c8044)

I tried above suggestions, however, no luck.

**Comment by dizonb** [comment]
I generate a migration with this code, but wonder why Datatype for enums are UNKNOWN. I'm using HeidiSQL instead of pgadmin4.

```
export enum ArticleType {
  NEWS = 'news',
  NEWS_RELEASE = 'news_release',
  COLUMN = 'column',
  CASE = 'case',
  OTHER = 'other',
}

export enum ArticleStatus {
  DRAFT = 'draft',
  PUBLISHED = 'published',
  ARCHIVED = 'archived',
}
```


```
@Column({
    type: 'enum',
    enum: ArticleType,
    default: ArticleType.OTHER,
  })
  articleType: ArticleType;

@Column({
    type: 'enum',
    enum: ArticleStatus,
    default: ArticleStatus.DRAFT,
  })
  status: ArticleStatus;
```

![Image](https://github.com/user-attachments/assets/f9fdda95-e43c-4f94-81dc-89bc8c7c8044)

I tried above suggestions, however, no luck.
